File: reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    Map<Integer, List<List<Integer>>> map = new HashMap<>();
    int count = 0;

    public int minReorder(int n, int[][] connections) {
        for (int[] connection : connections) {
            map.computeIfAbsent(connection[0], k -> new ArrayList<List<Integer>>()).add(Arrays.asList(connection[1], 1));
            map.computeIfAbsent(connection[1], k -> new ArrayList<List<Integer>>()).add(Arrays.asList(connection[0], 0));
        }
        dfs(0, -1);
        return count;
    }

    private void dfs(int node, int origin) {
        for (List<Integer> adj : map.get(node)) {
            int destination = adj.get(0);
            int direction = adj.get(1);
            if (destination != origin) {
                count += direction;
                dfs(destination, node);
            }
        }
    }
}

File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        adj = {}
        changes = 0
        visited = [0] * n
        for conn in connections:
            a, b  = conn
            if a not in adj:
                adj[a] = set()
            if b not in adj:
                adj[b] = set()
            adj[b].add(a)
            adj[a].add(-b)
        def dfs(city):
            nonlocal changes
            visited[city] = 1
            for neighbor in adj[city]:
                if neighbor < 0 and not visited[-neighbor]:
                    changes += 1
                    neighbor = -neighbor
                if not visited[abs(neighbor)]:
                    dfs(neighbor)
        dfs(0)
        return changes